src/runner.py

```diff
@@ -0,0 +1,576 @@
+collection of `Runner` instances.
+"""
+
+from __future__ import annotations
+
+import json
+import logging
+import time
+from pathlib import Path
+from typing import Iterable, Optional, Sequence
+
+from errors import LxdError, RunnerCreateError, RunnerError, RunnerFileLoadError, RunnerRemoveError
+from lxd import LxdInstance
+from lxd_type import LxdInstanceConfig
+from runner_type import (
+    GitHubOrg,
+    GitHubRepo,
+    RunnerClients,
+    RunnerConfig,
+    RunnerStatus,
+    VirtualMachineResources,
+)
+from utilities import retry
+
+logger = logging.getLogger(__name__)
+
+
+class Runner:
+    """Single instance of GitHub self-hosted runner.
+
+    Attrs:
+        app_name (str): Name of the charm.
+        path (GitHubPath): Path to GitHub repo or org.
+        proxies (ProxySetting): HTTP proxy setting for juju charm.
+        name (str): Name of the runner instance.
+        exist (bool): Whether the runner instance exists on LXD.
+        online (bool): Whether GitHub marks this runner as online.
+        busy (bool): Whether GitHub marks this runner as busy.
+    """
+
+    runner_application = Path("/opt/github-runner")
+    env_file = runner_application / ".env"
+    config_script = runner_application / "config.sh"
+    runner_script = runner_application / "start.sh"
+    pre_job_script = runner_application / "pre-job.sh"
+
+    def __init__(
+        self,
+        clients: RunnerClients,
+        runner_config: RunnerConfig,
+        runner_status: RunnerStatus,
+        instance: Optional[LxdInstance] = None,
+    ):
+        """Construct the runner instance.
+
+        Args:
+            clients: Clients to access various services.
+            runner_config: Configuration of the runner instance.
+            instance: LXD instance of the runner if already created.
+        """
+        # Dependency injection to share the instances across different `Runner` instance.
+        self._clients = clients
+        self.config = runner_config
+        self.status = runner_status
+        self.instance = instance
+
+    def create(
+        self,
+        image: str,
+        resources: VirtualMachineResources,
+        binary_path: Path,
+        registration_token: str,
+    ):
+        """Create the runner instance on LXD and register it on GitHub.
+
+        Args:
+            image: Name of the image to launch the LXD instance with.
+            resources: Resource setting for the LXD instance.
+            binary_path: Path to the runner binary.
+            registration_token: Token for registering the runner on GitHub.
+
+        Raises:
+            RunnerCreateError: Unable to create a LXD instance for runner.
+        """
+        logger.info("Creating runner: %s", self.config.name)
+
+        try:
+            self.instance = self._create_instance(image, resources)
+            self._start_instance()
+            # Wait some initial time for the instance to boot up
+            time.sleep(60)
```

Comment:
Shouldn't this be configurable and have a default value too?

Reply:
Minor comments